dask/bag/core.py

```diff
@@ -2498,3 +2478,121 @@ def to_dataframe(seq, columns, dtypes):
         seq = list(seq)
     res = pd.DataFrame(seq, columns=list(columns))
     return res.astype(dtypes, copy=False)
+
+
+def repartition_npartitions(bag, npartitions):
+    """ Changes the number of partitions of the bag.
+
+    This can be used to reduce or increase the number of partitions
+    of the bag.
+    """
+    new_name = "repartition-%d-%s" % (npartitions, tokenize(bag, npartitions))
+    if npartitions == bag.npartitions:
+        return bag
+    elif bag.npartitions > npartitions:
+        ratio = bag.npartitions / npartitions
+        new_partitions_boundaries = [
+            int(old_partition_index * ratio)
+            for old_partition_index in range(npartitions + 1)
+        ]
+        return _repartition_from_boundaries(bag, new_partitions_boundaries, new_name)
+    else:  # npartitions > bag.npartitions
+        div, mod = divmod(npartitions, bag.npartitions)
+        nsplits = [div] * bag.npartitions
+        nsplits[-1] += mod
+        return _split_partitions(bag, nsplits, new_name)
+
+
+def repartition_size(bag, size):
```

Comment:
This seems to share a lot of code with `dask.dataframe.core.repartition_size`. Do you see opportunities for deduplication?

Reply:
Yeah that's a good point. At first glance it looks like they could share most of the code in that method. The only differences are:
- total_mem_usage
- _split_partitions (the guts of which are the same but the way the return value is built differs, this could be easily handled)
- the df code uses pandas Series & numpy in different places whereas bag uses functions from toolz...that could probably be handled easily as well.

So I think a way to deduplicate would be to move the repartition_size method to a shared space (is `base.py` the place for this?). The reason I didn't start w/that is that it seems like it might be awkward to have that one method be shared while it seems like most of the rest of the operations are not. Let me know what you think.

Reply:
@TomAugspurger definitely there is duplicated code b/t bag and dataframe w.r.t. repartitioning. I just am not sure of the best way to make them share code in the existing project structure. 

It feels odd to put the shared code in the `DaskMethodsMixin` (which is already shared b/t bag and dataframe) as that seems to house very generic methods. Putting it in `utils`, which is also already shared, doesn't seem quite right either since those seem like general purpose methods not tied to needing a dask object. 

So I guess a third option would be to put the various partitioning methods in a separate utils file, called `partition_utils.py` or something like that, which only df and bag import. Does that sound like a good path forward, or am I missing some other way to share the code?

Reply:
FWIW, I'm comfortable with some code duplication here.  The complexity of factoring this out might be more than the complexity of duplication given that this is only going to happen twice.